win: add gettime for monotonic clock

wlemkows opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1282)
<!-- Reviewable:end -->

lplewa commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/windows/win_file.c, line 142 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTo2MDL9tAYhzNpvFqU:-KTo2MDL9tAYhzNpvFqV:byi1zai) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L142)):*
> ```C
> 
> void
> clock_gettime(int id, struct timespec *ts)
> ```

i'm not sure if win_file is a correct flie for this function

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/windows/win_file.c, line 141 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KToqxpzZrztLjgrIUq7:-KToqxpzZrztLjgrIUq8:b-wgb7on) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L141)):*
> ```C
> }
> 
> void
> ```

int as return value

---

*[src/windows/win_file.c, line 142 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTo2MDL9tAYhzNpvFqU:-KTor7XregeE3xZ2oaav:btjlee5) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L142)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> i'm not sure if win_file is a correct flie for this function
</details>
I suggest win_time.c

---

*[src/windows/win_file.c, line 166 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTor-y-Q0XGJOjvcZM0:-KTor-y-Q0XGJOjvcZM1:b-eczel6) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L166)):*
> ```C
> 			ts->tv_nsec = milisecond *  NANOSEC_IN_MILISEC;
> 		}
> 	}
> ```

return error if id is invalid (EINVAL) or not implemented (ENOSYS)

---

*[src/windows/include/platform.h, line 178 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KToqtv2-z-8dAOgtlHu:-KToqtv3mKF4AP3OWo7K:b-fnmqe7) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/include/platform.h#L178)):*
> ```objc
> #define NANOSEC_IN_MILISEC 1000000
> 
> void clock_gettime(int id, struct timespec *ts);
> ```

int clock_gettime(int id, struct timespec *ts); ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/windows/include/platform.h, line 178 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KToqtv2-z-8dAOgtlHu:-KTraWmS4WiyP5NDVMXW:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/include/platform.h#L178)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> int clock_gettime(int id, struct timespec *ts); ?
</details>
Done.

---

*[src/windows/win_file.c, line 141 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KToqxpzZrztLjgrIUq7:-KTraQ7YbOlhHgEWyylJ:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L141)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> int as return value
</details>
Done.

---

*[src/windows/win_file.c, line 142 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTo2MDL9tAYhzNpvFqU:-KTraQtnL_hY2rkIq_ME:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L142)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> I suggest win_time.c
</details>
Done.

---

*[src/windows/win_file.c, line 166 at r1](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTor-y-Q0XGJOjvcZM0:-KTraS_4QdJ32ymGNqZB:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/50d0871f6329c1449ee1442f663271d0ea836b8a/src/windows/win_file.c#L166)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> return error if id is invalid (EINVAL) or not implemented (ENOSYS)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/windows/win_time.c, line 34 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsTFFl_ZEEhDcGQrdB:-KTsTFFl_ZEEhDcGQrdC:b729ki2) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L34)):*
> ```C
> 
> /*
>  * win_timee.c -- Windows emulation of Linux-specific time functions
> ```

win_time.c

---

*[src/windows/win_time.c, line 37 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsT6pbZAmGboch-Xw_:-KTsT6pbZAmGboch-Xwa:b-6oleat) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L37)):*
> ```C
>  */
> 
> int
> ```

missing brief

---

*[src/windows/win_time.c, line 40 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsgbb5j02JWPuobh9u:-KTsgbb5j02JWPuobh9v:bs24bub) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L40)):*
> ```C
> clock_gettime(int id, struct timespec *ts)
> {
> 	if (id != CLOCK_MONOTONIC) {
> ```

So, CLOCK_REALTIME is defined, but not supported?

---

*[src/windows/win_time.c, line 47 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsY9Yq0n2J4R7DKfBA:-KTsY9Yq0n2J4R7DKfBB:b-dcrr2d) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L47)):*
> ```C
> 	if (id == CLOCK_MONOTONIC) {
> 
> 		unsigned long elapsed_time_milisec;
> ```

long long

---

*[src/windows/win_time.c, line 54 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsXdRmwHREkSxW9hiR:-KTsXdRnoyBLOvir5q-J:b-hr9nts) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L54)):*
> ```C
> 		 * that have elapsed since the system was started
> 		 */
> 		elapsed_time_milisec = GetTickCount();
> ```

GetTickCount64()

---

*[src/windows/win_time.c, line 55 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsg9vSKybyZ_4tOu50:-KTsg9vTk72Zql3WN8vr:b-cayqag) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L55)):*
> ```C
> 		 */
> 		elapsed_time_milisec = GetTickCount();
> 		if (elapsed_time_milisec % MILISEC_IN_SEC == 0) {
> ```

Do we need this huge if()? Isn't the last 'else' enough?

---

*[src/windows/include/platform.h, line 175 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsVfjHLIT7EuMvgKoe:-KTsVfjHLIT7EuMvgKof:b-gmjhl3) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/include/platform.h#L175)):*
> ```objc
> #define CLOCK_REALTIME 2
> 
> #define MILISEC_IN_SEC 1000
> ```

Are these macros required here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[src/windows/win_time.c, line 2 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTxKToImKicYVYmd7mT:-KTxKToJjamcYuMXQvtn:bb6ht2g) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L2)):*
> ```C
> /*
>  * Copyright 2015-2016, Intel Corporation
> ```

2016

---

*[src/windows/win_time.c, line 45 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTxKbiblfmtyn6FpKbd:-KTxKbiblfmtyn6FpKbe:b-6o59rj) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L45)):*
> ```C
> 	}
> 
> 	if (id == CLOCK_MONOTONIC) {
> ```

this if is always true 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Review status: 2 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/windows/win_time.c, line 2 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTxKToImKicYVYmd7mT:-KTxTLcpepR4Lwa4oRVE:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L2)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> 2016
</details>
Done.

---

*[src/windows/win_time.c, line 34 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsTFFl_ZEEhDcGQrdB:-KTxTMs6sRmUDXCxWOtP:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> win_time.c
</details>
Done.

---

*[src/windows/win_time.c, line 37 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsT6pbZAmGboch-Xw_:-KTxTN_dQwDWpzUV7-1c:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L37)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> missing brief
</details>
Done.

---

*[src/windows/win_time.c, line 40 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsgbb5j02JWPuobh9u:-KTxCtzr-04D9MMb-LLj:b-eafz9x) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L40)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> So, CLOCK_REALTIME is defined, but not supported?
</details>
I will remove #define CLOCK_REALTIME from my pull request. Implementation of that is here https://github.com/pmem/nvml/pull/1277/files in 	JLLLinn pull reqest.

---

*[src/windows/win_time.c, line 45 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTxKbiblfmtyn6FpKbd:-KTxT_oarH2K2uQSJqNN:b-2eti5q) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L45)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> this if is always true 
</details>
Done. Changed to switch case.

---

*[src/windows/win_time.c, line 47 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsY9Yq0n2J4R7DKfBA:-KTxTdXfmZn48QJ0o4Ug:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L47)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> long long
</details>
Done.

---

*[src/windows/win_time.c, line 54 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsXdRmwHREkSxW9hiR:-KTxTdz747tqnWpeHZOM:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L54)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> GetTickCount64()
</details>
Done.

---

*[src/windows/win_time.c, line 55 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsg9vSKybyZ_4tOu50:-KTxTeoJAni8E4s2o1We:b-4m07dp) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/win_time.c#L55)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Do we need this huge if()? Isn't the last 'else' enough?
</details>
No we don't :) Done.

---

*[src/windows/include/platform.h, line 175 at r2](https://reviewable.io:443/reviews/pmem/nvml/1282#-KTsVfjHLIT7EuMvgKoe:-KTxTt41aun69Vom3lG1:b-1mhwpn) ([raw file](https://github.com/pmem/nvml/blob/55ca0f197b81a3270746e08979e15027500ae071/src/windows/include/platform.h#L175)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Are these macros required here?
</details>
Done. These macros are helpful, I moved it to win_time

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282)*
<!-- Sent from Reviewable.io -->

plebioda commented:
You need to rebase your patch - there are merge conflicts.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/windows/win_time.c, line 48 at r3](https://reviewable.io:443/reviews/pmem/nvml/1282#-KU2_3Yt6UGaoiaWwBE8:-KU2_3YutKb_La9gu55Q:b-dfniut) ([raw file](https://github.com/pmem/nvml/blob/045f142578ff4d9b49028b64525c0d765e5948fc/src/windows/win_time.c#L48)):*
> ```C
> {
> 	switch (id) {
> 	case CLOCK_MONOTONIC:;
> ```

; after : ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1282#-:-KU2_BI1d0UMj_hpYLK7:b-myyhty)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Merged by hand.
